pkg/testutils/testcluster/testcluster.go

```diff
@@ -301,19 +303,54 @@ func (tc *TestCluster) Start(t testing.TB) {
 		errCh = make(chan error, nodes)
 	}
 
+	// Determine if we should probabilistically start a SQL server for the
+	// cluster.
+	const probabilityOfStartingTestTenant = 0.5
+	probabilisticallyStartTestSQLServer := true
+	startedTestSQLServer := true
+	if rand.Float32() > probabilityOfStartingTestTenant {
+		probabilisticallyStartTestSQLServer = false
+	}
+
 	disableLBS := false
 	for i := 0; i < nodes; i++ {
 		// Disable LBS if any server has a very low scan interval.
 		if tc.serverArgs[i].ScanInterval > 0 && tc.serverArgs[i].ScanInterval <= 100*time.Millisecond {
 			disableLBS = true
 		}
 
+		// If we're not probabilistically starting the test SQL server, disable
+		// its start and set the "started" flag accordingly. We need to do this
+		// with two separate if checks because the DisableDefaultSQLServer flag
+		// could have been set coming into this function by the caller.
+		if !probabilisticallyStartTestSQLServer {
+			tc.Servers[i].Cfg.DisableDefaultSQLServer = true
+		}
+		if tc.Servers[i].Cfg.DisableDefaultSQLServer {
+			if startedTestSQLServer && i > 0 {
+				t.Fatal(errors.Newf("starting only some nodes with a test SQL server is not"+
+					"currently supported - attempted to disable SQL sever on node %d", i))
+			}
+			startedTestSQLServer = false
+		}
+
 		if tc.clusterArgs.ParallelStart {
 			go func(i int) {
 				errCh <- tc.startServer(i, tc.serverArgs[i])
 			}(i)
 		} else {
 			if err := tc.startServer(i, tc.serverArgs[i]); err != nil {
+				if strings.Contains(err.Error(), "requires a CCL binary") {
+					if i != 0 {
+						t.Fatal(errors.Newf("failed to start server on node %d due to lack of "+
+							"CCL binary after server started successfully on previous nodes", i))
+					}
+					for j := 0; j < nodes; j++ {
+						tc.Servers[j].Stopper().Stop(context.TODO())
+					}
+					tc.Stopper().Stop(context.TODO())
+					skip.IgnoreLint(t, "skipping due to lack of CCL binary")
```

Comment:
This is a bit odd. Can't we prevent testcluster from going the tenant route if this is not a CCL binary, by having the ccl hook set some package variable that testcluster can read? That seems to be preferable to this rather complex logic that can lead to tests getting unexpectedly skipped a fraction of the time (if I read it correctly).

Reply:
It would also prevent all of this string matching to boot :-)